agentstack/frameworks/crewai.py

```diff
@@ -0,0 +1,281 @@
+from agentstack.tasks import TaskConfig
+from agentstack.agents import AgentConfig
+from agentstack.generation import asttools
+
+
+ENTRYPOINT: Path = Path('src/crew.py')
+
+
+class CrewFile(asttools.File):
+    """
+    Parses and manipulates the CrewAI entrypoint file.
+    All AST interactions should happen within the methods of this class.
+    """
+
+    _base_class: Optional[ast.ClassDef] = None
+
+    def get_base_class(self) -> ast.ClassDef:
+        """A base class is a class decorated with `@CrewBase`."""
+        if self._base_class is None:  # Gets cached to save repeat iteration
+            try:
+                self._base_class = asttools.find_class_with_decorator(self.tree, 'CrewBase')[0]
+            except IndexError:
+                raise ValidationError(f"`@CrewBase` decorated class not found in {ENTRYPOINT}")
+        return self._base_class
+
+    def get_crew_method(self) -> ast.FunctionDef:
+        """A `crew` method is a method decorated with `@crew`."""
+        try:
+            base_class = self.get_base_class()
+            return asttools.find_decorated_method_in_class(base_class, 'crew')[0]
+        except IndexError:
+            raise ValidationError(
+                f"`@crew` decorated method not found in `{base_class.name}` class in {ENTRYPOINT}"
+            )
+
+    def get_task_methods(self) -> list[ast.FunctionDef]:
+        """A `task` method is a method decorated with `@task`."""
+        return asttools.find_decorated_method_in_class(self.get_base_class(), 'task')
+
+    def add_task_method(self, task: TaskConfig):
+        """Add a new task method to the CrewAI entrypoint."""
+        task_methods = self.get_task_methods()
+        if task.name in [method.name for method in task_methods]:
+            # TODO this should check all methods in the class for duplicates
+            raise ValidationError(f"Task `{task.name}` already exists in {ENTRYPOINT}")
+        if task_methods:
+            # Add after the existing task methods
+            _, pos = self.get_node_range(task_methods[-1])
+        else:
+            # Add before the `crew` method
+            crew_method = self.get_crew_method()
+            pos, _ = self.get_node_range(crew_method)
+
+        code = f"""    @task
+    def {task.name}(self) -> Task:
+        return Task(
+            config=self.tasks_config['{task.name}'],
+        )"""
+        if not self.source[:pos].endswith('\n'):
+            code = '\n\n' + code
+        if not self.source[pos:].startswith('\n'):
+            code += '\n\n'
+        self.edit_node_range(pos, pos, code)
+
+    def get_agent_methods(self) -> list[ast.FunctionDef]:
+        """An `agent` method is a method decorated with `@agent`."""
+        return asttools.find_decorated_method_in_class(self.get_base_class(), 'agent')
+
+    def add_agent_method(self, agent: AgentConfig):
+        """Add a new agent method to the CrewAI entrypoint."""
+        # TODO do we want to pre-populate any tools?
```

Comment:
i think we just leave tools blank, unless they start with a template of course

Reply:
Templates add it later using `add_tool` right?